### PR TITLE
Add support for driver pool, instance flexibility policy, and min_num_instances for Dataproc

### PR DESCRIPTION
**Pull request by ahzaz**
---
Add following to ClusterGenerator for Dataproc
- min_num_workers: The minimum number of primary worker instances to create
- driver_pool_size: The size of the cluster driver pool.
- driver_pool_id: Custom identifier for the DRIVER Node Group being created.
- secondary_worker_instance_flexibility_policy: Instance flexibility Policy allowing a mixture of VM shapes and provisioning models.
- Upgrade google-cloud-dataproc to 5.5.0

### CODE REVIEW

File: airflow/providers/google/cloud/operators/dataproc.py
Comment (SamWheating) [inline airflow/providers/google/cloud/operators/dataproc.py:81]: Could these classes be simplified as dataclasses? Airflow's minimum supported python is 3.8 so it should be safe to use them here.

File: airflow/providers/google/cloud/operators/dataproc.py
Comment (SamWheating) [inline airflow/providers/google/cloud/operators/dataproc.py]:
```suggestion
    def _build_driver_pool(self):
```

File: airflow/providers/google/cloud/operators/dataproc.py
Comment (SamWheating) [inline airflow/providers/google/cloud/operators/dataproc.py]:
```suggestion
            cluster_data["auxiliary_node_groups"] = [self._build_driver_pool()]
```

**Review by SamWheating — COMMENTED** [review]
A couple suggestions

File: airflow/providers/google/provider.yaml
Comment (eladkal) [inline airflow/providers/google/provider.yaml:105]: Is there a reason why 5.4.x is not compatible with the changes added in this PR?

File: airflow/providers/google/cloud/operators/dataproc.py
Comment (ahzaz) [reply airflow/providers/google/cloud/operators/dataproc.py:81]: Done

File: airflow/providers/google/provider.yaml
Comment (ahzaz) [reply airflow/providers/google/provider.yaml:105]: The new fields (min_num_instances and instance_flexibility_policy) are available from 5.5.0

Comment (uranusjr) [reply airflow/providers/google/provider.yaml:105]: Would it be possible to detect the version and only allow them in 5.5.0, and keep compatibility to the lower version? (Is it worthwhile?)

Comment (ahzaz) [reply airflow/providers/google/provider.yaml:105]: It is possible to detect the version using `__version__` but don't think it would be a good idea to do so.

Comment (uranusjr) [reply airflow/providers/google/provider.yaml:105]: Can you elaborate?

Comment (potiuk) [reply airflow/providers/google/provider.yaml:105]: > Would it be possible to detect the version and only allow them in 5.5.0, and keep compatibility to the lower version? (Is it worthwhile?)

I think not worthwile -> generally the "cloud" services work best with newest versions of libraries, and from the point of view of Airlfow "google provider" is the only consumer of this library. 

Generally speaking - when we are talking about providers, as long as we already use higher version of the library in main constraints (we do https://github.com/apache/airflow/blob/constraints-main/constraints-3.10.txt#L321C1-L322C1)  - bumping minimum version of pretty much any dependency (except some really popular and possibly not easy to upgrade to) is safe. Users can always use earlier version of provider if they require the dependency in lower version, and actually in most installation cases this will happen automatically - provider will be downgraded in case user will want to install non-constraint compatible version of the dependency.

So pretty much in any case where we have a reason to bump min version of dependeny in a provider, we should do it (unless we have a very good reason not to - but this should be exceptional case).

Comment (ahzaz) [reply airflow/providers/google/provider.yaml:105]: Agree with above. Also, can someone please trigger the remaining workflows.

**Comment by ahzaz** [comment]
@potiuk @SamWheating Can you please review/approve the PR if there are no more concerns ?

File: airflow/providers/google/cloud/operators/dataproc.py
Comment (SamWheating) [inline airflow/providers/google/cloud/operators/dataproc.py:120]: Its not immediately clear how this value interacts with `num_workers`, might be worth elaborating a bit more?

From the API docs:
> Example: Cluster creation request with num_instances = 5 and min_num_instances = 3:
>
> If 4 VMs are created and 1 instance fails, the failed VM is deleted. The cluster is resized to 4 instances and placed in a RUNNING state.
> If 2 instances are created and 3 instances fail, the cluster in placed in an ERROR state. The failed VMs are not deleted.

Could we add some of that context here? I don't have any strong opinions on the exact wording.

File: airflow/providers/google/cloud/operators/dataproc.py
Comment (SamWheating) [inline airflow/providers/google/cloud/operators/dataproc.py:418]: Should we raise an Exception at initialization time if `min_num_workers > num_workers`?

It's usually best to surface issues as early as possible.

File: airflow/providers/google/cloud/operators/dataproc.py
Comment (SamWheating) [inline airflow/providers/google/cloud/operators/dataproc.py:373]: Thoughts on allowing instance type here as well? Something like:

```suggestion
                "node_group_config": {
                    "num_instances": self.driver_pool_size
                    "machine_type_uri": self.driver_pool_machine_type
                 },
```

Maybe this is too much config though? I'm fine with leaving this for another day.

**Review by SamWheating — COMMENTED** [review]
A couple comments / suggestions

File: airflow/providers/google/cloud/operators/dataproc.py
Comment (ahzaz) [reply airflow/providers/google/cloud/operators/dataproc.py:373]: Let keep it these two for now. Later on, we can driver pool config (like InstanceFlexibilityPolicy) to cover rest of the config options.

File: airflow/providers/google/cloud/operators/dataproc.py
Comment (ahzaz) [reply airflow/providers/google/cloud/operators/dataproc.py:418]: Done

File: airflow/providers/google/cloud/operators/dataproc.py
Comment (ahzaz) [reply airflow/providers/google/cloud/operators/dataproc.py:120]: Updated.

**Review by SamWheating — APPROVED** [review]

**Comment by ahzaz** [comment]
@SamWheating Thank you for review. Can you please trigger the workflow?

**Comment by SamWheating** [comment]
I actually can't trigger the workflow - we'll have to wait for a committer to come along and do the final check + merge.

**Comment by ahzaz** [comment]
@potiuk Can you please help get this merged?

**Review by Taragolis — APPROVED** [review]
Looks good.

**Comment by ahzaz** [comment]
@Taragolis Can you please help merge this.

**Review by potiuk — APPROVED** [review]

**Pull request merged**
